### PR TITLE
PEP 604: Remove some more uses of Union/Optional

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Mypy seems fine with these. Aliases that use type[] or variadic tuple[]
still break mypy.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
Some of these mypy is cool with on >= 3.10, but can't handle on <3.10

**Comment by JelleZijlstra** [comment]
Right. Looks like these are the patterns:
- `type[]` at top level fails
- `tuple[T1, T2]` at top level fails (but `tuple[T1, ...]` is fine)
- `T1 | Callable[..., T2 | T3]` fails, but only <=3.9

Still ended up with a few fixes. I didn't look at `@python2`.

File: stdlib/xmlrpc/client.pyi [hunk 1]
```diff
@@ -13,9 +13,11 @@ class _SupportsTimeTuple(Protocol):
...
 
 _DateTimeComparable = DateTime | datetime | str | _SupportsTimeTuple
-_Marshallable = Union[None, bool, int, float, str, bytes, tuple[Any, ...], list[Any], dict[Any, Any], datetime, DateTime, Binary]
-_XMLDate = Union[int, datetime, tuple[int, ...], time.struct_time]
-_HostType = Union[tuple[str, dict[str, str]], str]
+_Marshallable = (
+    bool | int | float | str | bytes | None | tuple[Any, ...] | list[Any] | dict[Any, Any] | datetime | DateTime | Binary
+)
+_XMLDate = int | datetime | tuple[int, ...] | time.struct_time
+_HostType = Union[str, tuple[str, dict[str, str]]]
```
Comment (AlexWaygood) [inline stdlib/xmlrpc/client.pyi:20]: Why has this one changed?

Comment (JelleZijlstra) [reply stdlib/xmlrpc/client.pyi:20]: I guess I moved the order around to see if it made a difference, then reverted back to Union wrong
```suggestion
_HostType = Union[tuple[str, dict[str, str]], str]
```

**Comment by AlexWaygood** [comment]
> Right. Looks like these are the patterns:
> 
> * `type[]` at top level fails
> * `tuple[T1, T2]` at top level fails (but `tuple[T1, ...]` is fine)
> * `T1 | Callable[..., T2 | T3]` fails, but only <=3.9
> 
> Still ended up with a few fixes. I didn't look at `@python2`.

Of these, as far as I'm aware, only (2) is tracked over at mypy (in https://github.com/python/mypy/issues/11098); (1) and (3) need to have bug reports filed.

**Review by AlexWaygood — APPROVED** [review]
LGTM

**Comment by JelleZijlstra** [comment]
I created https://github.com/python/mypy/labels/topic-pep-604 so we can track these better. Going to spend some time today organizing the mypy issue tracker.

**Comment by AlexWaygood** [comment]
> I created https://github.com/python/mypy/labels/topic-pep-604 so we can track these better. Going to spend some time today organizing the mypy issue tracker.

https://github.com/python/mypy/issues/11887 could do with that label being added 👍

**Comment by AlexWaygood** [comment]
Mypy bugs for (1) and (3) have been filled:

- https://github.com/python/mypy/issues/12392
- https://github.com/python/mypy/issues/12393

**Comment by JelleZijlstra** [comment]
Thanks!

**Pull request merged**